File: src/main/java/com/fineos/idam/petInsurance/pet/api/PetsApi.java
/*---------------------------------------------------------------------------------------------
FINEOS

(c) Copyright FINEOS.
ALL RIGHTS RESERVED

---------------------------------------------------------------------------------------------*/
package com.fineos.idam.petInsurance.pet.api;

import com.fineos.idam.petInsurance.pet.dto.Error;
import com.fineos.idam.petInsurance.pet.dto.Pet;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.enums.ParameterIn;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.annotation.Generated;
import jakarta.validation.Valid;
import java.util.List;
import java.util.Optional;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.context.request.NativeWebRequest;

@Generated("v1.0.0")
@Validated
@Tag(name = "pets", description = "the pets API")
@SuppressWarnings("PMD.AvoidDuplicateLiterals")
public interface PetsApi {

    /**
     * Returns an {@link Optional} containing the current {@link NativeWebRequest}.
     *
     * <p>By default, this implementation returns an empty {@link Optional}. Subclasses may override this method to
     * provide a non-empty {@link Optional} containing the actual {@link NativeWebRequest} if available.
     *
     * @return an {@link Optional} containing the current {@link NativeWebRequest} if available, otherwise an empty
     *     {@link Optional}
     */
    default Optional<NativeWebRequest> getRequest() {
        return Optional.empty();
    }

    /**
     * POST /pets : Create a pet. Create a pet in the store with the given details.
     *
     * @param pet Pet object that needs to be added to the store. (required)
     * @return Null response. (status code 201) or unexpected error, (status code 200)
     */
    @Operation(
            operationId = "createPets",
            summary = "Create a pet.",
            description = "Create a pet in the store with the given details.",
            tags = {"pets"},
            responses = {
                @ApiResponse(responseCode = "201", description = "Null response."),
                @ApiResponse(
                        responseCode = "default",
                        description = "unexpected error,",
                        content = {
                            @Content(
                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    schema = @Schema(implementation = Error.class))
                        })
            })
    @PostMapping(
            value = "/pets",
            produces = {"application/json"},
            consumes = {"application/json"})
    default ResponseEntity<Void> createPets(
            @Parameter(name = "Pet", description = "Pet object that needs to be added to the store.", required = true)
                    @Valid
                    @RequestBody
                    Pet pet) {
        return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
    }

    /**
     * GET /pets : List all pets. Get a paged list of pets.
     *
     * @param limit How many items to return at one time (max 100). (optional)
     * @return A paged array of pets. (status code 200) or unexpected error. (status code 200)
     */
    @Operation(
            operationId = "listPets",
            summary = "List all pets.",
            description = "Get a paged list of pets.",
            tags = {"pets"},
            responses = {
                @ApiResponse(
                        responseCode = "200",
                        description = "A paged array of pets.",
                        content = {
                            @Content(
                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    array = @ArraySchema(schema = @Schema(implementation = Pet.class)))
                        }),
                @ApiResponse(
                        responseCode = "default",
                        description = "unexpected error.",
                        content = {
                            @Content(
                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    schema = @Schema(implementation = Error.class))
                        })
            })
    @GetMapping(
            value = "/pets",
            produces = {"application/json"})
    default ResponseEntity<List<Pet>> listPets(
            @Parameter(
                            name = "limit",
                            description = "How many items to return at one time (max 100).",
                            in = ParameterIn.QUERY)
                    @Valid
                    @RequestParam(required = false)
                    Integer limit) {
        return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
    }

    /**
     * GET /pets/{petId} : Info for a specific pet. Get a specific pet by ID.
     *
     * @param petId The id of the pet to retrieve. (required)
     * @return Expected response to a valid request. (status code 200) or unexpected error. (status code 200)
     */
    @Operation(
            operationId = "showPetById",
            summary = "Info for a specific pet.",
            description = "Get a specific pet by ID.",
            tags = {"pets"},
            responses = {
                @ApiResponse(
                        responseCode = "200",
                        description = "Expected response to a valid request.",
                        content = {
                            @Content(
                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    schema = @Schema(implementation = Pet.class))
                        }),
                @ApiResponse(
                        responseCode = "default",
                        description = "unexpected error.",
                        content = {
                            @Content(
                                    mediaType = MediaType.APPLICATION_JSON_VALUE,
                                    schema = @Schema(implementation = Error.class))
                        })
            })
    @GetMapping(
            value = "/pets/{petId}",
            produces = {"application/json"})
    default ResponseEntity<Pet> showPetById(
            @Parameter(
                            name = "petId",
                            description = "The id of the pet to retrieve.",
                            required = true,
                            in = ParameterIn.PATH)
                    @PathVariable
                    String petId) {
        return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
    }
}

File: src/com/fineos/idam/petinsurance/pet/api/pets_api.py
# coding: utf-8

from typing import Dict, List  # noqa: F401
import importlib
import pkgutil

from com.fineos.idam.petinsurance.pet.api.pets_api_base import BasePetsApi
import com.fineos.idam.petinsurance.pet.impl

from fastapi import (  # noqa: F401
    APIRouter,
    Body,
    Cookie,
    Depends,
    Form,
    Header,
    HTTPException,
    Path,
    Query,
    Response,
    Security,
    status,
)

from com.fineos.idam.petinsurance.pet.dto.extra_models import TokenModel  # noqa: F401
from com.fineos.idam.petinsurance.pet.dto.error import Error
from com.fineos.idam.petinsurance.pet.dto.pet import Pet


router = APIRouter()

ns_pkg = com.fineos.idam.petinsurance.pet.impl
for _, name, _ in pkgutil.iter_modules(ns_pkg.__path__, ns_pkg.__name__ + "."):
    importlib.import_module(name)


@router.post(
    "/pets",
    responses={
        201: {"description": "Null response."},
        200: {"model": Error, "description": "unexpected error,"},
    },
    tags=["pets"],
    summary="Create a pet.",
    response_model_by_alias=True,
)
async def create_pets(
    pet: Pet = Body(None, description="Pet object that needs to be added to the store."),
) -> None:
    """Create a pet in the store with the given details."""
    if not BasePetsApi.subclasses:
        raise HTTPException(status_code=501, detail="Not implemented")

    if len(BasePetsApi.subclasses) > 1:
        raise HTTPException(status_code=500, detail="Multiple implementations found, You can have only one implementation for the base class")
            
    return await BasePetsApi.subclasses[0]().create_pets(pet)


@router.get(
    "/pets",
    responses={
        200: {"model": List[Pet], "description": "A paged array of pets."},
        200: {"model": Error, "description": "unexpected error."},
    },
    tags=["pets"],
    summary="List all pets.",
    response_model_by_alias=True,
)
async def list_pets(
    limit: int = Query(None, description="How many items to return at one time (max 100).", alias="limit"),
) -> List[Pet]:
    """Get a paged list of pets."""
    if not BasePetsApi.subclasses:
        raise HTTPException(status_code=501, detail="Not implemented")

    if len(BasePetsApi.subclasses) > 1:
        raise HTTPException(status_code=500, detail="Multiple implementations found, You can have only one implementation for the base class")
            
    return await BasePetsApi.subclasses[0]().list_pets(limit)


@router.get(
    "/pets/{petId}",
    responses={
        200: {"model": Pet, "description": "Expected response to a valid request."},
        200: {"model": Error, "description": "unexpected error."},
    },
    tags=["pets"],
    summary="Info for a specific pet.",
    response_model_by_alias=True,
)
async def show_pet_by_id(
    petId: str = Path(..., description="The id of the pet to retrieve."),
) -> Pet:
    """Get a specific pet by ID."""
    if not BasePetsApi.subclasses:
        raise HTTPException(status_code=501, detail="Not implemented")

    if len(BasePetsApi.subclasses) > 1:
        raise HTTPException(status_code=500, detail="Multiple implementations found, You can have only one implementation for the base class")
            
    return await BasePetsApi.subclasses[0]().show_pet_by_id(petId)
